client/header/utils.js

```diff
@@ -0,0 +1,99 @@
+/**
+ * External dependencies
+ */
+import { Slot, Fill } from '@wordpress/components';
+import { cloneElement } from '@wordpress/element';
+
+/**
+ * Ordered header item.
+ *
+ * @param {Node} children - Node children.
+ * @param {number} order - Node order.
+ * @param {Array} props - Fill props.
+ * @return {Node} Node.
+ */
+const createOrderedChildren = ( children, order, props ) => {
+	return typeof children === 'function'
+		? cloneElement( children( props ), { order } )
+		: cloneElement( children, { ...props, order } );
+};
+
+/**
+ * Create a Fill for extensions to add items to the WooCommerce Admin header.
+ *
+ * @slotFill WooHeaderItem
+ * @example
+ * const MyHeaderItem = () => (
+ * <WooHeaderItem>My header item</WooHeaderItem>
+ * );
+ *
+ * registerPlugin( 'my-extension', {
+ * render: MyHeaderItem,
+ * scope: 'woocommerce-admin',
+ * } );
+ * @param {Object} param0
+ * @param {Array} param0.children - Node children.
+ * @param {Array} param0.order - Node order.
+ */
+export const WooHeaderItem = ( { children, order = 1 } ) => {
```

Comment:
I wonder if we should just create a `WooHeaderItemBefore` and `WooHeaderItemAfter` with an order instead of using the negative order values.
The negative vs positive seems a little hacky, and creating two separate HeaderItems might be easier.

Reply:
I'm a little mixed on this mostly due to the naming.  "Before" and "After" make me think "before/after what?"   The answer to that is `WooHeaderItemBeforePageTitle` but that's quite long and not very pleasing to read.

`WooHeaderItemLeft` might be slightly better, but let me give this some more thought.

Reply:
>I would steer away from "Left" or "Right", because they are unclear when a right-to-left layout is used (such as Arabic and Hebrew).

Great point, @mattsherman.

>WooHeaderItemBeforePageTitle certainly is wordy, but it is pretty clear.

I agree that it's more clear than Before/After, though I wonder if something with more intent may be beneficial here.

Maybe  `WooHeaderItem` is retained for the header items to the right and `WooHeaderNavigation` is the area to the left which is primarily for navigational use.  Any thoughts on this @mattsherman @louwie17?

Reply:
> Maybe WooHeaderItem is retained for the header items to the right and WooHeaderNavigation is the area to the left which is primarily for navigational use

I quite like that approach and naming +1 for me.

Reply:
Oh, I like that... `WooHeaderNavigation` +1